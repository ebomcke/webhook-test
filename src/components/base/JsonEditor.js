import React from 'react';
import brace from 'brace';
import 'brace/mode/json';
import 'brace/theme/github';
import AceEditor from 'react-ace';

const JsonEditor = ({ height, value, onChange }) => (
  <AceEditor
    mode="json"
    theme="github"
    wrapEnabled={true}
    width="100%"
    height={height}
    value={value}
    onChange={onChange}
  />
);

export default JsonEditor;