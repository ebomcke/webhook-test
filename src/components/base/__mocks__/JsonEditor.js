import React from 'react';

const JsonEditor = props => (
  <div>
    This is a placeholder for the JSON editor.
    <pre>
      <code>{JSON.stringify(props)}</code>
    </pre>
  </div>
);

export default JsonEditor;