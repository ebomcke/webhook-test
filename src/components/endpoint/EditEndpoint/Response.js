import React, { Suspense } from 'react';
import { Form, Message, Icon, Loader, Dimmer, Segment } from 'semantic-ui-react';
const JsonEditor = React.lazy(() =>
  import(/* webpackChunkName: "JsonEditor" */ '../../base/JsonEditor'),
);

const Response = ({ endpointEditor, nextButton, renderMessage }) => (
  <div>
    {!!renderMessage && (
      <Message icon>
        <Icon name="help" />
        <Message.Content>
          <Message.Header>This is optional</Message.Header>
          <p>
            If you're not sure what the default response should be you can
            safely leave this default value.
          </p>
        </Message.Content>
      </Message>
    )}
    <Form>
      <Form.Field>
        <Form.Select
          fluid
          name="defaultResponseStatusCode"
          onChange={(event, data) =>
            endpointEditor.onStatusCodeChange(data.value)
          }
          label="Status code"
          options={[
            { key: 1, text: '200 - OK', value: 200 },
            { key: 2, text: '201 - Created', value: 201 },
          ]}
          value={endpointEditor.endpoint.defaultResponseStatusCode}
        />
      </Form.Field>
      <Form.Field>
        <label>Response body</label>
        <Suspense
          fallback={
            <Segment>
              <Dimmer active inverted>
                <Loader size="big" inverted>
                  Loading
                </Loader>
              </Dimmer>
            </Segment>
          }
        >
          <JsonEditor
            height="300px"
            value={endpointEditor.endpoint.defaultResponseBody}
            onChange={endpointEditor.onResponseBodyChange}
          />
        </Suspense>
      </Form.Field>
      {nextButton()}
    </Form>
  </div>
);

export default Response;
