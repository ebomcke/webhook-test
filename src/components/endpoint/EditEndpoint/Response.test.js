import React from 'react';
import { create } from 'react-test-renderer';
import Response from './Response';
import JsonEditor from '../../base/JsonEditor';

const endpointEditor = {
  onStatusCodeChange: jest.fn(),
  onResponseBodyChange: jest.fn(),
  endpoint: {
    defaultResponseStatusCode: 200,
    defaultResponseBody: JSON.stringify({
      status: 'OK',
    }),
  },
};

const nextButton = () => <div>Next button placeholder</div>;

describe('Rendering tests', () => {
  test('Without info message', () => {
    const response = create(
      <Response
        endpointEditor={endpointEditor}
        nextButton={nextButton}
        renderMessage={false}
      />,
    );
    expect(response.toJSON()).toMatchSnapshot();
  });

  test('With info message', () => {
    const response = create(
      <Response
        endpointEditor={endpointEditor}
        nextButton={nextButton}
        renderMessage={true}
      />,
    );
    expect(response.toJSON()).toMatchSnapshot();
  });
});

describe('Interactions tests', () => {
  const response = create(
    <Response
      endpointEditor={endpointEditor}
      nextButton={nextButton}
      renderMessage={true}
    />,
  );

  afterEach(() => {
    endpointEditor.onStatusCodeChange.mockClear();
    endpointEditor.onResponseBodyChange.mockClear();
  });

  test('Changing the response status code updates the endpoint editor', () => {
    const statusCodeDropdown = response.root.findByProps({
      name: 'defaultResponseStatusCode',
    });
    statusCodeDropdown.props.onChange(null, {
      value: 201,
    });
    expect(endpointEditor.onStatusCodeChange.mock.calls.length).toBe(1);
    expect(endpointEditor.onStatusCodeChange.mock.calls[0][0]).toEqual(201);
  });

  test('Changing the response body updates the endpoint editor', () => {
    const jsonEditor = response.root.findByType(JsonEditor);
    const newBody = JSON.stringify({ from: 'unit-test' });
    jsonEditor.props.onChange(newBody);
    expect(endpointEditor.onResponseBodyChange.mock.calls.length).toBe(1);
    expect(endpointEditor.onResponseBodyChange.mock.calls[0][0]).toEqual(
      newBody,
    );
  });
});
