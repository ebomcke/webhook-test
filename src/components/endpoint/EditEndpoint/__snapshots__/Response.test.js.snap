// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rendering tests With info message 1`] = `
<div>
  <div
    className="ui icon message"
  >
    <i
      aria-hidden="true"
      className="help icon"
    />
    <div
      className="content"
    >
      <div
        className="header"
      >
        This is optional
      </div>
      <p>
        If you're not sure what the default response should be you can safely leave this default value.
      </p>
    </div>
  </div>
  <form
    className="ui form"
    onSubmit={[Function]}
  >
    <div
      className="field"
    >
      <div
        className="field"
      >
        <label>
          Status code
        </label>
        <div
          aria-expanded={false}
          className="ui fluid selection dropdown"
          name="defaultResponseStatusCode"
          onBlur={[Function]}
          onChange={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onMouseDown={[Function]}
          role="listbox"
          tabIndex={0}
        >
          <div
            aria-live="polite"
            className="text"
            role="alert"
          >
            200 - OK
          </div>
          <i
            aria-hidden="true"
            className="dropdown icon"
            onClick={[Function]}
          />
          <div
            className="menu transition"
          >
            <div
              aria-checked={true}
              aria-selected={true}
              className="active selected item"
              onClick={[Function]}
              role="option"
              style={
                Object {
                  "pointerEvents": "all",
                }
              }
            >
              <span
                className="text"
              >
                200 - OK
              </span>
            </div>
            <div
              aria-checked={false}
              aria-selected={false}
              className="item"
              onClick={[Function]}
              role="option"
              style={
                Object {
                  "pointerEvents": "all",
                }
              }
            >
              <span
                className="text"
              >
                201 - Created
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      className="field"
    >
      <label>
        Response body
      </label>
      <div>
        This is a placeholder for the JSON editor.
        <pre>
          <code>
            {"height":"300px","value":"{\\"status\\":\\"OK\\"}"}
          </code>
        </pre>
      </div>
    </div>
    <div>
      Next button placeholder
    </div>
  </form>
</div>
`;

exports[`Rendering tests Without info message 1`] = `
<div>
  <form
    className="ui form"
    onSubmit={[Function]}
  >
    <div
      className="field"
    >
      <div
        className="field"
      >
        <label>
          Status code
        </label>
        <div
          aria-expanded={false}
          className="ui fluid selection dropdown"
          name="defaultResponseStatusCode"
          onBlur={[Function]}
          onChange={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onMouseDown={[Function]}
          role="listbox"
          tabIndex={0}
        >
          <div
            aria-live="polite"
            className="text"
            role="alert"
          >
            200 - OK
          </div>
          <i
            aria-hidden="true"
            className="dropdown icon"
            onClick={[Function]}
          />
          <div
            className="menu transition"
          >
            <div
              aria-checked={true}
              aria-selected={true}
              className="active selected item"
              onClick={[Function]}
              role="option"
              style={
                Object {
                  "pointerEvents": "all",
                }
              }
            >
              <span
                className="text"
              >
                200 - OK
              </span>
            </div>
            <div
              aria-checked={false}
              aria-selected={false}
              className="item"
              onClick={[Function]}
              role="option"
              style={
                Object {
                  "pointerEvents": "all",
                }
              }
            >
              <span
                className="text"
              >
                201 - Created
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      className="field"
    >
      <label>
        Response body
      </label>
      <div>
        This is a placeholder for the JSON editor.
        <pre>
          <code>
            {"height":"300px","value":"{\\"status\\":\\"OK\\"}"}
          </code>
        </pre>
      </div>
    </div>
    <div>
      Next button placeholder
    </div>
  </form>
</div>
`;
